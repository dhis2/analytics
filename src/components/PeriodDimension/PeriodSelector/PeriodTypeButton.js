import React, { Component } from 'react'
import PropTypes from 'prop-types'
import i18n from '@dhis2/d2-i18n'

import styles from './styles/PeriodTypeButton.style'

class PeriodTypeButton extends Component {
    handleClick = () => {
        this.props.onClick(this.props.periodType)
    }

    render = () => (
        <>
            <button
                className={`nav-button ${
                    this.props.periodType === this.props.activePeriodType
                        ? 'active'
                        : ''
                }`}
                onClick={this.handleClick}
            >
                {i18n.t(this.props.text)}
            </button>
            <style jsx>{styles}</style>
        </>
    )
}

PeriodTypeButton.propTypes = {
    activePeriodType: PropTypes.string.isRequired,
    periodType: PropTypes.string.isRequired,
    text: PropTypes.string.isRequired,
    onClick: PropTypes.func.isRequired,
}

export default PeriodTypeButton
