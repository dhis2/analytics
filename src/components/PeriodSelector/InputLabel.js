import React from 'react'
import PropTypes from 'prop-types'
import MuiLabel from '@material-ui/core/InputLabel'
import { withStyles } from '@material-ui/core/styles'

const styles = {
    inputLabel: {
        color: '#595959',
        whiteSpace: 'nowrap',
    },
}

const InputLabel = ({ classes, label }) => {
    return (
        <>
            <MuiLabel className={classes.inputLabel} htmlFor="period-type">
                {label}
            </MuiLabel>
        </>
    )
}

InputLabel.propTypes = {
    label: PropTypes.string.isRequired,
    classes: PropTypes.object,
}

export default withStyles(styles)(InputLabel)
