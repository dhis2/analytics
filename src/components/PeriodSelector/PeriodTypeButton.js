import React, { Component } from 'react'
import PropTypes from 'prop-types'

import styles from './styles/PeriodTypeButton.style'

class PeriodTypeButton extends Component {
    handleClick = () => {
        this.props.onClick(this.props.periodType)
    }

    render = () => (
        <>
            <button
                className={`nav-button ${
                    this.props.periodType === this.props.activePeriodType
                        ? 'active'
                        : ''
                }`}
                onClick={this.handleClick}
            >
                {this.props.text}
            </button>
            <style jsx>{styles}</style>
        </>
    )
}

PeriodTypeButton.propTypes = {
    activePeriodType: PropTypes.string.isRequired,
    periodType: PropTypes.string.isRequired,
    text: PropTypes.string.isRequired,
    onClick: PropTypes.func.isRequired,
}

export default PeriodTypeButton
