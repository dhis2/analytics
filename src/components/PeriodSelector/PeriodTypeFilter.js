import React from 'react'
import PropTypes from 'prop-types'

import FormControl from '@material-ui/core/FormControl'
import Select from '@material-ui/core/Select'
import MenuItem from '@material-ui/core/MenuItem'
import i18n from '@dhis2/d2-i18n'
import InputLabel from './InputLabel'

const PeriodTypeFilter = ({ options, onChange, value }) => {
    return (
        <FormControl>
            <InputLabel label={i18n.t('Period type')} />
            <Select
                onChange={onChange}
                value={value}
                inputProps={{ name: 'periodType', id: 'period-type' }}
                disableUnderline
            >
                {Object.entries(options).map(([key, val]) => (
                    <MenuItem value={key} key={key}>
                        {val.name()}
                    </MenuItem>
                ))}
            </Select>
        </FormControl>
    )
}

PeriodTypeFilter.propTypes = {
    options: PropTypes.object.isRequired,
    value: PropTypes.string,
    onChange: PropTypes.func,
}

export default PeriodTypeFilter
