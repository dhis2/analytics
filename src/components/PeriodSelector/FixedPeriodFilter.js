import React, { Component } from 'react'
import PropTypes from 'prop-types'

import InputLabel from './InputLabel'
import Select from '@material-ui/core/Select'
import MenuItem from '@material-ui/core/MenuItem'
import Menu from '@material-ui/core/Menu'
import FormControl from '@material-ui/core/FormControl'
import ArrowDownIcon from '@material-ui/icons/KeyboardArrowDown'
import ArrowUpIcon from '@material-ui/icons/KeyboardArrowUp'

import i18n from '@dhis2/d2-i18n'
import PeriodTypeFilter from './PeriodTypeFilter'
import FixedPeriodsGenerator, { MONTHLY } from './modules/FixedPeriodsGenerator'

export const defaultState = {
    periodType: MONTHLY,
    year: new Date().getFullYear(),
    yearsOffset: 0,
    yearSelectElement: null,
}

export const YEARS_RANGE = 8

class FixedPeriodFilter extends Component {
    constructor(props) {
        super(props)

        this.periodsGenerator = new FixedPeriodsGenerator()
        this.state = defaultState
    }

    componentDidMount = () => {
        const periods = this.generatePeriods(
            this.state.periodType,
            this.state.year
        )

        this.props.setOfferedPeriods(periods, true)
    }

    onPeriodTypeChange = event => {
        this.setState({
            periodType: event.target.value,
        })

        if (this.state.year) {
            this.props.setOfferedPeriods(
                this.generatePeriods(event.target.value, this.state.year)
            )
        }
    }

    onYearChange = event => {
        this.setState({
            year: event.target.value,
            yearSelectElement: null,
        })

        if (this.state.periodType) {
            this.props.setOfferedPeriods(
                this.generatePeriods(this.state.periodType, event.target.value)
            )
        }
    }

    onYearSelectClick = event => {
        this.setState({ yearSelectElement: event.currentTarget })
    }

    getYears = () => {
        let years = []

        years = years.concat(
            [
                ...Array(
                    Math.floor(YEARS_RANGE / 2) +
                        (YEARS_RANGE % 2 === 0 ? 1 : 2)
                ).keys(),
            ]
                .slice(1)
                .reverse()
                .map(
                    offset =>
                        new Date().getFullYear() -
                        offset +
                        this.state.yearsOffset
                )
        )

        years = years.concat(
            [...Array(Math.floor(YEARS_RANGE / 2)).keys()].map(
                offset =>
                    new Date().getFullYear() + offset + this.state.yearsOffset
            )
        )

        return years
    }

    generatePeriods = (periodType, year) => {
        const generator = this.periodsGenerator.get(periodType)

        return generator
            .generatePeriods({
                offset: year - new Date().getFullYear(),
                filterFuturePeriods: false,
                reversePeriods: false,
            })
            .map((period, idx) => ({ ...period, idx }))
    }

    closeYearSelect = () => {
        this.setState({ yearSelectElement: null })
    }

    shiftYearsBack = () => {
        this.setState({ yearsOffset: this.state.yearsOffset - YEARS_RANGE })
    }

    shiftYearsForth = () => {
        this.setState({ yearsOffset: this.state.yearsOffset + YEARS_RANGE })
    }

    renderYearSelectValue = () => this.state.year

    render() {
        const years = this.getYears()

        return (
            <>
                <PeriodTypeFilter
                    options={this.periodsGenerator.getOptions()}
                    value={this.state.periodType}
                    onChange={this.onPeriodTypeChange}
                />
                <FormControl>
                    <InputLabel label={i18n.t('Year')} />
                    <Select
                        SelectDisplayProps={{
                            style: { cursor: 'pointer', color: '#000000' },
                            onClick: this.onYearSelectClick,
                        }}
                        value={this.state.year}
                        inputProps={{ name: 'year', id: 'year' }}
                        renderValue={this.renderYearSelectValue}
                        disableUnderline
                        disabled
                    />
                    <Menu
                        MenuListProps={{
                            style: { width: '130px' },
                        }}
                        anchorEl={this.state.yearSelectElement}
                        open={Boolean(this.state.yearSelectElement)}
                        onClose={this.closeYearSelect}
                    >
                        <MenuItem
                            value=""
                            key="shiftYearsBack"
                            onClick={this.shiftYearsBack}
                        >
                            <ArrowUpIcon />
                        </MenuItem>
                        {years.map(year => (
                            <MenuItem
                                onClick={this.onYearChange}
                                key={year}
                                value={year}
                                selected={this.state.year === year}
                            >
                                {year}
                            </MenuItem>
                        ))}
                        <MenuItem
                            value=""
                            key="shiftYearsForth"
                            onClick={this.shiftYearsForth}
                        >
                            <ArrowDownIcon />
                        </MenuItem>
                    </Menu>
                </FormControl>
            </>
        )
    }
}

FixedPeriodFilter.propTypes = {
    setOfferedPeriods: PropTypes.func.isRequired,
}

export default FixedPeriodFilter
