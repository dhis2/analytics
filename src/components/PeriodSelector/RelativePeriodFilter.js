import React, { Component } from 'react'
import PropTypes from 'prop-types'

import PeriodTypeFilter from './PeriodTypeFilter'
import RelativePeriodsGenerator, {
    MONTHS,
} from './modules/RelativePeriodsGenerator'

export const defaultState = {
    periodType: MONTHS,
}

class RelativePeriods extends Component {
    constructor(props) {
        super(props)

        this.state = defaultState
        this.periodsGenerator = new RelativePeriodsGenerator()
    }

    componentDidMount = () => {
        const periods = this.generatePeriods(this.state.periodType)
        this.props.setOfferedPeriods(periods, true)
    }

    onPeriodTypeChange = event => {
        this.setState({
            periodType: event.target.value,
        })

        this.props.setOfferedPeriods(this.generatePeriods(event.target.value))
    }

    generatePeriods = periodType => {
        const generator = this.periodsGenerator.get(periodType)

        return generator
            .generatePeriods()
            .map((period, idx) => ({ ...period, idx }))
    }

    render = () => {
        return (
            <PeriodTypeFilter
                options={this.periodsGenerator.getOptions()}
                value={this.state.periodType}
                onChange={this.onPeriodTypeChange}
            />
        )
    }
}

RelativePeriods.propTypes = {
    setOfferedPeriods: PropTypes.func.isRequired,
}

export default RelativePeriods
