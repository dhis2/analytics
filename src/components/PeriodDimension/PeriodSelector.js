import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { Transfer, TabBar, Tab } from '@dhis2/ui-core'
import i18n from '@dhis2/d2-i18n'

import FixedPeriodFilter from './FixedPeriodFilter'
import RelativePeriodFilter from './RelativePeriodFilter'
import { MONTHS, getRelativePeriodsOptionsById } from './utils/relativePeriods'
import { MONTHLY, getFixedPeriodsOptionsById } from './utils/fixedPeriods'
import styles from '../styles/DimensionSelector.style'
import { TransferOption } from '../TransferOption'
import PeriodIcon from '../../assets/DimensionItemIcons/PeriodIcon'
import {
    TRANSFER_HEIGHT,
    TRANSFER_OPTIONS_WIDTH,
    TRANSFER_SELECTED_WIDTH,
} from '../../modules/dimensionSelectorHelper'

const defaultRelativePeriodType = getRelativePeriodsOptionsById(MONTHS)
const defaultFixedPeriodType = getFixedPeriodsOptionsById(MONTHLY)
const defaultFixedPeriodYear = new Date().getFullYear()
const fixedPeriodConfig = year => ({
    offset: year - defaultFixedPeriodYear,
    filterFuturePeriods: false,
    reversePeriods: false,
})

class PeriodSelector extends Component {
    state = {
        allPeriods: defaultRelativePeriodType.getPeriods(),
        selectedPeriods: [],
        isRelative: true,
        relativeFilter: {
            periodType: {
                label: defaultRelativePeriodType.getName(),
                value: defaultRelativePeriodType.id,
            },
        },
        fixedFilter: {
            periodType: {
                label: defaultFixedPeriodType.getName(),
                value: defaultFixedPeriodType.id,
            },
            year: defaultFixedPeriodYear.toString(),
        },
    }

    constructor(props) {
        super(props)

        this.state.selectedPeriods = this.props.initialSelectedPeriods.map(
            item => ({
                label: item.name,
                value: item.id,
                key: item.id,
            })
        )
    }

    onIsRelativeClick = isRelative => {
        if (this.state.isRelative !== isRelative) {
            this.setState({ isRelative })
            this.setState({
                allPeriods: isRelative
                    ? getRelativePeriodsOptionsById(
                          this.state.relativeFilter.periodType.value
                      ).getPeriods()
                    : getFixedPeriodsOptionsById(
                          this.state.fixedFilter.periodType.value
                      ).getPeriods(
                          fixedPeriodConfig(Number(this.state.fixedFilter.year))
                      ),
            })
        }
    }

    renderHeader = () => (
        <>
            <TabBar>
                <Tab
                    selected={this.state.isRelative}
                    onClick={() => this.onIsRelativeClick(true)}
                >
                    {i18n.t('Relative periods')}
                </Tab>
                <Tab
                    selected={!this.state.isRelative}
                    onClick={() => this.onIsRelativeClick(false)}
                >
                    {i18n.t('Fixed periods')}
                </Tab>
            </TabBar>
            <div className="filterContainer">
                {this.state.isRelative ? (
                    <RelativePeriodFilter
                        currentFilter={this.state.relativeFilter.periodType}
                        onSelectFilter={filter => {
                            this.setState({
                                relativeFilter: { periodType: filter },
                            })
                            this.setState({
                                allPeriods: getRelativePeriodsOptionsById(
                                    filter.value
                                ).getPeriods(),
                            })
                        }}
                    />
                ) : (
                    <FixedPeriodFilter
                        currentPeriodType={this.state.fixedFilter.periodType}
                        currentYear={this.state.fixedFilter.year}
                        onSelectPeriodType={periodType => {
                            this.onSelectFixedPeriods({
                                periodType,
                                year: this.state.fixedFilter.year,
                            })
                        }}
                        onSelectYear={year => {
                            this.onSelectFixedPeriods({
                                periodType: this.state.fixedFilter.periodType,
                                year,
                            })
                        }}
                    />
                )}
            </div>
            <style jsx>{styles}</style>
        </>
    )

    onSelectFixedPeriods = fixedFilter => {
        this.setState({
            fixedFilter,
            allPeriods: getFixedPeriodsOptionsById(
                fixedFilter.periodType.value
            ).getPeriods(fixedPeriodConfig(Number(fixedFilter.year))),
        })
    }

    renderEmptySelection = () => (
        <>
            <p className="emptySelection">{i18n.t('No periods selected')}</p>
            <style jsx>{styles}</style>
        </>
    )

    render = () => (
        <Transfer
            onChange={({ selected }) => {
                this.setState({ selectedPeriods: selected })
                this.props.onSelect(selected)
            }}
            selected={this.state.selectedPeriods}
            leftHeader={this.renderHeader()}
            enableOrderChange
            height={TRANSFER_HEIGHT}
            optionsWidth={TRANSFER_OPTIONS_WIDTH}
            selectedWidth={TRANSFER_SELECTED_WIDTH}
            selectedEmptyComponent={this.renderEmptySelection()}
            rightFooter={this.props.rightFooter}
            // TODO: Add rightHeader "Selected Periods" once the Transfer component supports this (https://github.com/dhis2/ui-core/issues/885)
        >
            {this.state.allPeriods.map(item => (
                <TransferOption
                    label={item.getName()}
                    value={item.id}
                    key={item.id}
                    icon={PeriodIcon}
                />
            ))}
        </Transfer>
    )
}

PeriodSelector.propTypes = {
    onSelect: PropTypes.func.isRequired,
    initialSelectedPeriods: PropTypes.arrayOf(PropTypes.object),
    rightFooter: PropTypes.node,
}

PeriodSelector.defaultProps = {
    initialSelectedPeriods: [],
}

export default PeriodSelector
