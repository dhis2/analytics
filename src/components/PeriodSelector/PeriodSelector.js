import React, { Component } from 'react'
import PropTypes from 'prop-types'
import i18n from '@dhis2/d2-i18n'

import ItemSelector from '../ItemSelector/ItemSelector'
import PeriodTypeButton from './PeriodTypeButton'
import FixedPeriodFilter from './FixedPeriodFilter'
import RelativePeriodFilter from './RelativePeriodFilter'
import { FIXED, RELATIVE } from './modules/periodTypes'
import styles from './styles/PeriodSelector.style'

class PeriodSelector extends Component {
    state = {
        offeredPeriods: [],
        offeredPeriodsInOrder: [],
        selectedPeriods: [],
        periodType: RELATIVE,
    }

    constructor(props) {
        super(props)

        this.state.selectedPeriods = this.props.selectedItems
    }

    onPeriodTypeClick = periodType => {
        if (this.state.periodType !== periodType) {
            this.setState({ periodType })
        }
    }

    onSelectPeriods = periodIds => {
        const offeredPeriods = this.state.offeredPeriods.filter(
            period => !periodIds.includes(period.id)
        )
        const newPeriods = this.state.offeredPeriods.filter(period =>
            periodIds.includes(period.id)
        )
        const selectedPeriods = this.state.selectedPeriods.concat(newPeriods)

        this.setState({ selectedPeriods, offeredPeriods })
        this.props.onSelect(selectedPeriods)
    }

    setSelectedPeriodOrder = periodIds => {
        const selectedPeriods = periodIds.map(id =>
            this.state.selectedPeriods.find(period => period.id === id)
        )

        this.setState({ selectedPeriods })
        this.props.onReorder(selectedPeriods)
    }

    onDeselectPeriods = periodIds => {
        const selectedPeriods = this.state.selectedPeriods.filter(
            period => !periodIds.includes(period.id)
        )
        const removedPeriods = this.state.selectedPeriods.filter(period =>
            periodIds.includes(period.id)
        )
        const offeredPeriods = this.state.offeredPeriodsInOrder.filter(
            period => !selectedPeriods.map(p => p.id).includes(period.id)
        )

        this.setState({ selectedPeriods, offeredPeriods })
        this.props.onDeselect(removedPeriods)
    }

    initializeOfferedPeriods = (periods, initial = false) => {
        const selectedPeriods = initial
            ? this.props.selectedItems
            : this.state.selectedPeriods
        const offeredPeriods = periods.filter(
            period => !selectedPeriods.map(p => p.id).includes(period.id)
        )

        this.setState({ offeredPeriodsInOrder: periods, offeredPeriods })
    }

    render = () => {
        const FilterOptions =
            this.state.periodType === FIXED
                ? FixedPeriodFilter
                : RelativePeriodFilter

        const unselected = {
            items: this.state.offeredPeriods,
            onSelect: this.onSelectPeriods,
            filterText: '',
        }

        const selected = {
            items: this.state.selectedPeriods,
            onDeselect: this.onDeselectPeriods,
            onReorder: this.setSelectedPeriodOrder,
        }

        return (
            <>
                <div>
                    <PeriodTypeButton
                        active={this.state.periodType === RELATIVE}
                        text={i18n.t('Relative periods')}
                        onClick={() => this.onPeriodTypeClick(RELATIVE)}
                    />
                    <PeriodTypeButton
                        active={this.state.periodType === FIXED}
                        text={i18n.t('Fixed periods')}
                        onClick={() => this.onPeriodTypeClick(FIXED)}
                    />
                </div>
                <div className="period-selector">
                    <ItemSelector unselected={unselected} selected={selected}>
                        <div className="options-area">
                            <FilterOptions
                                setOfferedPeriods={
                                    this.initializeOfferedPeriods
                                }
                            />
                        </div>
                    </ItemSelector>
                </div>
                <style jsx>{styles}</style>
            </>
        )
    }
}

PeriodSelector.propTypes = {
    onDeselect: PropTypes.func.isRequired,
    onReorder: PropTypes.func.isRequired,
    onSelect: PropTypes.func.isRequired,
    selectedItems: PropTypes.arrayOf(PropTypes.object),
}

PeriodSelector.defaultProps = {
    selectedItems: [],
}

export default PeriodSelector
