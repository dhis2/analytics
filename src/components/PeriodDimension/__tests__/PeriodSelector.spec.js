import React from 'react'
import { shallow } from 'enzyme'
import PeriodSelector from '../PeriodSelector'

describe('The Period Selector component', () => {
    let props
    let shallowPeriodSelector

    const getWrapper = () => {
        if (!shallowPeriodSelector) {
            shallowPeriodSelector = shallow(<PeriodSelector {...props} />)
        }
        return shallowPeriodSelector
    }

    beforeEach(() => {
        props = {
            initialSelectedPeriods: [],
            onSelect: jest.fn(),
            rightFooter: <></>,
        }
        shallowPeriodSelector = undefined
    })

    it('matches the snapshot', () => {
        const wrapper = getWrapper()

        expect(wrapper).toMatchSnapshot()
    })
})
