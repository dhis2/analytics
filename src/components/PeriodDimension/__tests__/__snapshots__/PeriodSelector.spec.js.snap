// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The Period Selector component matches the snapshot 1`] = `
<Transfer
  dataTest="dhis2-uicore-transfer"
  enableOrderChange={true}
  filterCallback={[Function]}
  filterCallbackPicked={[Function]}
  height="512px"
  initialSearchTerm=""
  initialSearchTermPicked=""
  leftHeader={
    <React.Fragment>
      <TabBar
        dataTest="dhis2-uicore-tabbar"
      >
        <Tab
          dataTest="dhis2-uicore-tab"
          onClick={[Function]}
          selected={true}
        >
          Relative periods
        </Tab>
        <Tab
          dataTest="dhis2-uicore-tab"
          onClick={[Function]}
          selected={false}
        >
          Fixed periods
        </Tab>
      </TabBar>
      <div
        className="filterContainer"
      >
        <RelativePeriodFilter
          currentFilter="Months"
          onSelectFilter={[Function]}
        />
      </div>
      <style>
        
      </style>
    </React.Fragment>
  }
  maxSelections={Infinity}
  onChange={[Function]}
  options={
    Array [
      Object {
        "label": "This month",
        "value": "THIS_MONTH",
      },
      Object {
        "label": "Last month",
        "value": "LAST_MONTH",
      },
      Object {
        "label": "Last 3 months",
        "value": "LAST_3_MONTHS",
      },
      Object {
        "label": "Last 6 months",
        "value": "LAST_6_MONTHS",
      },
      Object {
        "label": "Last 12 months",
        "value": "LAST_12_MONTHS",
      },
      Object {
        "label": "Months this year",
        "value": "MONTHS_THIS_YEAR",
      },
    ]
  }
  optionsWidth="420px"
  renderOption={[Function]}
  rightFooter={<React.Fragment />}
  rightHeader={
    <React.Fragment>
      <p
        className="rightHeader"
      >
        Selected Periods
      </p>
      <style>
        
      </style>
    </React.Fragment>
  }
  selected={Array []}
  selectedEmptyComponent={
    <React.Fragment>
      <p
        className="emptyList"
      >
        No periods selected
      </p>
      <style>
        
      </style>
    </React.Fragment>
  }
  selectedWidth="298px"
/>
`;
